crates/ra_assists/src/lib.rs

```diff
@@ -77,6 +77,55 @@ where
     })
 }
 
+/// A functionality for locating imports for the given name.
+///
+/// Currently has to be a trait with the real implementation provided by the ra_ide_api crate,
+/// due to the search functionality located there.
+/// Later, this trait should be removed completely and the search functionality moved to a separate crate,
+/// accessible from the ra_assists crate.
+pub trait ImportsLocator<'a> {
+    /// Finds all imports for the given name and the module that contains this name.
+    fn find_imports(
+        &mut self,
+        name_to_import: InFile<&NameRef>,
+        module_with_name_to_import: Module,
+    ) -> Option<Vec<ModPath>>;
```

Comment:
obviously it would be nicest if we didn't need this at all, but if we need it, maybe we can make the interface smaller? i.e. just directly wrapping the index lookup, a function that takes a `Name` and returns a vector of `ModuleDef`s, instead of also putting the path generation in there?

Reply:
I think this should be a very temporary trait before we decide how to properly extract the functionality we need into a separate crate.
But I like the idea in general, although I have a few questions:

* `classify_name_ref` requires `InFile<&ast::NameRef>`, `classify_name` requires `InFile<&ast::Name>`, so not sure how I can pass `Name` only there.
I'm also not sure how to get the `Name` in the first place, since for the case
```
fn main() {
    let zz = HasMap<|>::new();
}
```
there's no `NAME` around the coursor, only `NAME_REF`.
So it looks like I should require `InFile<&ast::NameRef>` as an input here.

* The biggest issue here are the tests: since I have to keep them in the `ra_assists`, I have to create a mock trait implementation. I can return _something_ as a valid `ModPath` there, but how can I create the `ModuleDef` in the `ra_assists`?
Moveover, I cannot get any random `ModuleDef`, I need only those that the `find_use_path` method will consider as impotable.
Is there a way to get one with those properties?

If not, maybe we should focus on the code extraction instead to remove these hacks.

Reply:
I think `classify_name_ref` should be called outside that function (I just now realize that that's where we check whether the name resolves); that's part of making it smaller. `classify_name` gets called on the import candidate, not the name we're searching for. I actually meant a `hir::Name`, not an `ast::Name`, though probably a `String` would be just fine (it's getting converted to one anyway). Conceptually, since this is just an index lookup, it shouldn't need a file location or AST node.

For tests, I think we could write a small implementation that just does a linear search through all defs in the crate by name? The advantage for tests would be that we'd be testing a bigger part of the functionality.

Reply:
As for `classify_name_ref`, I think you actually don't need 90% of what it does; since you already are on an `ast::Path`, `analyzer.resolve_path` should be enough.

Reply:
Have implemented all the suggestions, the interface looks way better now, thanks!

One thing that bothers me though: I had to reexport a `ModuleDefId` in since I could not find a better way to extract the `LocalModuleId` that I needed to implement the linear search through all defs in the crate for the test `impl ImportsLocator for TestImportsLocator`.